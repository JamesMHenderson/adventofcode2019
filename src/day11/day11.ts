import { IntcodeComputer } from '../intcodeComputer';

const positions = (input: string, shipStart: { [position: string]: number } = {}) => {
    let direction = 0;
    const currentPosition = [0, 0];

    const ship: { [position: string]: number } = shipStart;

    const changeDirection = (rotation: number) => {
        if (rotation === 0) {
            direction--;
            if (direction < 0) direction = 3;
        } else if (rotation === 1) {
            direction++;
            direction = direction % 4;
        } else {
            throw new Error('Error');
        }
    };

    const updatePosition = (rotation: number) => {
        changeDirection(rotation);

        switch (direction) {
            case 0:
                currentPosition[1]++;
                break;
            case 1:
                currentPosition[0]++;
                break;
            case 2:
                currentPosition[1]--;
                break;
            case 3:
                currentPosition[0]--;
                break;
            default:
                throw new Error('Invalid direction');
        }
    };

    const intcodeComputer = new IntcodeComputer(input);

    while (!intcodeComputer.finished) {
        const position = currentPosition.join(',');
        const currentColour = ship[position] || 0;
        intcodeComputer.addInput(currentColour);
        intcodeComputer.run();
        ship[position] = intcodeComputer.output;
        intcodeComputer.run();
        updatePosition(intcodeComputer.output);
    }

    return ship;
};

export { positions };
