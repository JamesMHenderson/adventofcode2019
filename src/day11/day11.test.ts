import { positions } from './day11';
import input from './input.txt';

describe('Day 11', () => {
    describe('Part 1', () => {
        test('challenge', () => {
            // not 5654
            expect(Object.keys(positions(input.trim())).length).toBe(1771);
        });

        test('challenge part 2', () => {
            const ship = positions(input.trim(), { '0,0': 1 });

            const shipPos = Object.keys(ship).map(position => position.split(',').map(i => parseInt(i)));

            const minX = shipPos.reduce((acc: number, value: Array<number>) => (value[0] < acc ? value[0] : acc), 0);
            const maxX = shipPos.reduce((acc: number, value: Array<number>) => (value[0] > acc ? value[0] : acc), 0);
            const minY = shipPos.reduce((acc: number, value: Array<number>) => (value[1] < acc ? value[1] : acc), 0);
            const maxY = shipPos.reduce((acc: number, value: Array<number>) => (value[1] > acc ? value[1] : acc), 0);

            const paintedShip = Array(maxY - minY + 1)
                .fill(' ')
                .map(() => [...Array(maxX - minX + 1).fill(' ')]);

            Object.keys(ship).forEach(key => {
                const pos = key.split(',').map(i => parseInt(i));
                paintedShip[pos[1] - minY][pos[0] - minX] = ship[key] === 0 ? ' ' : '#';
            });

            expect(paintedShip.map(row => row.join(''))).toMatchSnapshot();
        });
    });
});