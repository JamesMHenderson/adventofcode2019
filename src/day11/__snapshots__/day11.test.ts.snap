// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Day 11 Part 1 challenge part 2 1`] = `
Array [
  "            ##                                                                                                                                                                                                                       ",
  "           # ## #                                                                                                                                                                                                                    ",
  "          ##  #                 ##                                                                                                                                                                                                   ",
  " ##          ##                # ## #                                                                                                                                                                                                ",
  " ####   ### #                 ##  #                 ##                                                                                                                                                                               ",
  "#  # #  ##           ##          ##                # ## #                                                                                                                                                                            ",
  "  #    ## #          ####   ### #                 ##  #                 ##                                                                                                                                                           ",
  " # ## #####         #  # #  ##           ##          ##                # ## #                                                                                                                                                        ",
  "     #  ###    #      #### ## #          ####   ### #                 ##  #                 ##                                                                                                                                       ",
  " ###  #####  # ####  # ########         #  # #  ##           ##          ##                # ## #                                                                                                                                    ",
  " ### # #  ###  # ## ### ##  ###    #      #### ## #          ####   ### #                 ##  #                 ##                                                                                                                   ",
  " # ##    ##  ##       ##  #####  # ####  # ########         #  # #  ##           ##          ##                # ## #                                                                                                                ",
  " ###  #             # ## # #  ###  # ## ### ##  ###    #      #### ## #          ####   ### #                 ##  #                 ##                                                                                               ",
  "   #   # #    ###    # ##    ##  ##       ##  #####  # ####  # ########         #  # #  ##           ##          ##                # ## #                                                                                            ",
  "    #  # ##  ####  #####  #             # ## # #  ###  # ## ### ##  ###    #      #### ## #          ####   ### #                 ##  #                 ##                                                                           ",
  "    ####   ### # ### # #   # #    ###    # ##    ##  ##       ##  #####  # ####  # ########         #  # #  ##           ##          ##                # ## #                                                                        ",
  "                #  ##   #  # ##  ####  #####  #             # ## # #  ###  # ## ### ##  ###    #      #### ## #          ####   ### #                 ##  #                 ##                                                       ",
  "                        ####   ### # ### # #   # #    ###    # ##    ##  ##       ##  #####  # ####  # ########         #  # #  ##           ##          ##                # ## #                                                    ",
  "                                    #  ##   #  # ##  ####  #####  #             # ## # #  ###  # ## ### ##  ###    #      #### ## #          ####   ### #                 ##  #                 ##                                   ",
  "                                            ####   ### # ### # #   # #    ###    # ##    ##  ##       ##  #####  # ####  # ########         #  # #  ##           ##          ##                # ## #                                ",
  "                                                        #  ##   #  # ##  ####  #####  #             # ## # #  ###  # ## ### ##  ###    #      #### ## #          ####   ### #                 ##  #                 ##               ",
  "                                                                ####   ### # ### # #   # #    ###    # ##    ##  ##       ##  #####  # ####  # ########         #  # #  ##           ##          ##                # ## #            ",
  "                                                                            #  ##   #  # ##  ####  #####  #             # ## # #  ###  # ## ### ##  ###    #      #### ## #          ####   ### #                 ##  #              ",
  "                                                                                    ####   ### # ### # #   # #    ###    # ##    ##  ##       ##  #####  # ####  # ########         #  # #  ##           ##          ##              ",
  "                                                                                                #  ##   #  # ##  ####  #####  #             # ## # #  ###  # ## ### ##  ###    #      #### ## #          ####   ### #                ",
  "                                                                                                        ####   ### # ### # #   # #    ###    # ##    ##  ##       ##  #####  # ####  # ########         #  # #  ##                   ",
  "                                                                                                                    #  ##   #  # ##  ####  #####  #             # ## # #  ###  # ## ### ##  ###    #      #### ## #                  ",
  "                                                                                                                            ####   ### # ### # #   # #    ###    # ##    ##  ##       ##  #####  # ####  # ########                  ",
  "                                                                                                                                        #  ##   #  # ##  ####  #####  #             # ## # #  ###  # ## ### ##  ###    #             ",
  "                                                                                                                                                ####   ### # ### # #   # #    ###    # ##    ##  ##       ##  #####  # ####          ",
  "                                                                                                                                                            #  ##   #  # ##  ####  #####  #             # ## # #  ###  # ##          ",
  "                                                                                                                                                                    ####   ### # ### # #   # #    ###    # ##    ##  ##              ",
  "                                                                                                                                                                                #  ##   #  # ##  ####  #####  #                      ",
  "                                                                                                                                                                                        ####   ### # ### # #   # #    ###    # ##    ",
  "                                                                                                                                                                                                    #  ##   #  # ##  ####  #   ###   ",
  "                                                                                                                                                                                                            ####   ### # ###### ##   ",
  "                                                                                                                                                                                                                        #   # # # #  ",
  "                                                                                                                                                                                                                           # ### # ##",
  "                                                                                                                                                                                                                          # ###  #  #",
  "                                                                                                                                                                                                                          ## # ## ###",
  "                                                                                                                                                                                                                           # # # ####",
  "                                                                                                                                                                                  #  #  ### #### #  #  ##  #  #  ##  #####  ##     # ",
  "                                                                                                                                                                                  #  # #  # #    #  # #  # #  # #  # #   # ##   #    ",
  "                                                                                                                                                                                  #  # # ## #    #  #    # #  # #  #  #  ### #   #   ",
  "                                                                                                                                                                                  #### #    ###  ####    # #### #  #   #   #  #      ",
  "                                                                                                                                                                                  #  # #  # #    #  #    # #  # #  #    # #          ",
  "                                                                                                                                                                                  #  #  ##  #### #  #   ## #  # #  # ####            ",
]
`;
